File: destack-ts/src/encoder/kompakt/encoder.ts
import { getObjectKey, KOMPAKT_OBJECT_ENCODERS } from "@destack/encoder/kompakt/core";
import {
  BinaryReader,
  BinaryWriter,
  type BuiltinObject,
  type Encoder,
  NodeType,
  ObjectKind,
  type Session,
  StructType,
  type Type,
} from "@destack/language/core";

/** Encoder for our Kompakt format. */
export class KompaktEncoder implements Encoder<Uint8Array> {
  packObject(kind: ObjectKind, metatype: NodeType | StructType, object: BuiltinObject): Uint8Array {
    const encoder = KOMPAKT_OBJECT_ENCODERS[getObjectKey(kind, metatype)];
    if (!encoder) {
      throw new Error(
        `no KompaktObjectEncoder for ${ObjectKind[kind] ?? kind}:${NodeType[metatype] ?? StructType[metatype] ?? metatype}`,
      );
    }
    const writer = new BinaryWriter();
    encoder.packObject(object, writer);
    return writer.toBytes();
  }

  packObjectBinary(
    kind: ObjectKind,
    metatype: NodeType | StructType,
    object: BuiltinObject,
    writer: BinaryWriter,
  ): void {
    const encoder = KOMPAKT_OBJECT_ENCODERS[getObjectKey(kind, metatype)];
    if (!encoder) {
      throw new Error(
        `no KompaktObjectEncoder for ${ObjectKind[kind] ?? kind}:${NodeType[metatype] ?? StructType[metatype] ?? metatype}`,
      );
    }
    encoder.packObject(object, writer);
  }

  unpackObject(
    kind: ObjectKind,
    metatype: NodeType | StructType,
    value: Uint8Array,
    session: Session | null,
  ): BuiltinObject {
    const encoder = KOMPAKT_OBJECT_ENCODERS[getObjectKey(kind, metatype)];
    if (!encoder) {
      throw new Error(
        `no KompaktObjectEncoder for ${ObjectKind[kind] ?? kind}:${NodeType[metatype] ?? StructType[metatype] ?? metatype}`,
      );
    }
    const reader = new BinaryReader(value);
    return encoder.unpackObject(reader, session);
  }

  unpackObjectBinary(
    kind: ObjectKind,
    metatype: NodeType | StructType,
    reader: BinaryReader,
    session: Session | null,
  ): BuiltinObject {
    const encoder = KOMPAKT_OBJECT_ENCODERS[getObjectKey(kind, metatype)];
    if (!encoder) {
      throw new Error(
        `no KompaktObjectEncoder for ${ObjectKind[kind] ?? kind}:${NodeType[metatype] ?? StructType[metatype] ?? metatype}`,
      );
    }
    return encoder.unpackObject(reader, session);
  }

  packValue(value: any, type: Type): Uint8Array {
    throw new Error("not implemented");
  }

  packValueBytes(value: any, type: Type, writer: BinaryWriter): void {
    throw new Error("not implemented");
  }

  unpackValue(type: Type, value: Uint8Array, session: Session | null): any {
    throw new Error("not implemented");
  }

  unpackValueBytes(type: Type, reader: BinaryReader, session: Session | null): any {
    throw new Error("not implemented");
  }
}

File: destack/core/encoding/encoder.py
from typing import TYPE_CHECKING, Any, Optional

from ..builtin import (
    EncoderFlag,
    Handle,
    HandleType,
    Object,
    ObjectKind,
    UInt32,
    declare_handle,
    declare_method,
)
from .binary import BinaryReader, BinaryWriter

if TYPE_CHECKING:
    from destack import Session, Type


@declare_handle(HandleType.ENCODER, is_abstract=True)
class Encoder(Handle):
    """Encoder for packing/unpacking Objects."""

    @declare_method(100)
    def pack_object(
        self,
        object: Object,
        options: EncoderFlag = EncoderFlag.DEFAULT,
    ) -> Any:
        """Pack an Object into some encoded format."""
        raise NotImplementedError

    @declare_method(101)
    def unpack_object(
        self,
        kind: ObjectKind | None,
        type: UInt32 | None,
        value: Any,
        session: Optional["Session"],
        options: EncoderFlag = EncoderFlag.DEFAULT,
    ) -> Object:
        """
        Unpack an Object from some encoded format.
        If type is not provided, the ObjectKind and metatype will be inferred/consumed from the value.
        """
        raise NotImplementedError

    @declare_method(102)
    def pack_object_binary(
        self,
        object: Object,
        writer: "BinaryWriter",
        options: EncoderFlag = EncoderFlag.DEFAULT,
    ) -> None:
        """Pack an Object into the byte representation of its encoded format."""
        raise NotImplementedError

    @declare_method(103)
    def unpack_object_binary(
        self,
        kind: ObjectKind | None,
        type: UInt32 | None,
        reader: "BinaryReader",
        session: Optional["Session"],
        options: EncoderFlag = EncoderFlag.DEFAULT,
    ) -> Object:
        """
        Unpack an Object from the byte representation of its encoded format.
        If type is not provided, the ObjectKind and metatype will be inferred/consumed from the value.
        """
        raise NotImplementedError

    @declare_method(110)
    def pack_type(
        self,
        type: "Type",
        options: EncoderFlag = EncoderFlag.DEFAULT,
    ) -> Any:
        """Pack a Type into some encoded format."""
        raise NotImplementedError

    @declare_method(111)
    def unpack_type(
        self,
        value: Any,
        options: EncoderFlag = EncoderFlag.DEFAULT,
    ) -> Any:
        """Unpack a Type from some encoded format."""
        raise NotImplementedError

    @declare_method(112)
    def pack_type_binary(
        self,
        type: "Type",
        writer: "BinaryWriter",
        options: EncoderFlag = EncoderFlag.DEFAULT,
    ) -> None:
        """Pack a Type into the byte representation of its encoded format."""
        raise NotImplementedError

    @declare_method(113)
    def unpack_type_binary(
        self,
        reader: "BinaryReader",
        options: EncoderFlag = EncoderFlag.DEFAULT,
    ) -> Any:
        """Unpack a Type from the byte representation of its encoded format."""
        raise NotImplementedError

    @declare_method(120)
    def pack_value(
        self,
        type: "Type",
        value: Any,
        options: EncoderFlag = EncoderFlag.DEFAULT,
    ) -> Any:
        """Pack a value into some encoded format."""
        raise NotImplementedError

    @declare_method(121)
    def unpack_value(
        self,
        type: "Type",
        value: Any,
        session: Optional["Session"],
        options: EncoderFlag = EncoderFlag.DEFAULT,
    ) -> Any:
        """Unpack a value from some encoded format."""
        raise NotImplementedError

    @declare_method(122)
    def pack_value_binary(
        self,
        type: "Type",
        value: Any,
        writer: "BinaryWriter",
        options: EncoderFlag = EncoderFlag.DEFAULT,
    ) -> None:
        """Pack a value into the byte representation of its encoded format."""
        raise NotImplementedError

    @declare_method(123)
    def unpack_value_binary(
        self,
        type: "Type",
        reader: "BinaryReader",
        session: Optional["Session"],
        options: EncoderFlag = EncoderFlag.DEFAULT,
    ) -> Any:
        """Unpack a value from the byte representation of its encoded format."""
        raise NotImplementedError
